core/store/src/trie/state_snapshot.rs

```diff
@@ -7,14 +7,56 @@ use crate::{Store, StoreConfig};
 use near_primitives::block::Block;
 use near_primitives::errors::EpochError;
 use near_primitives::errors::StorageError;
-use near_primitives::errors::StorageError::StorageInconsistentState;
 use near_primitives::hash::CryptoHash;
 use near_primitives::shard_layout::ShardUId;
 
+use std::error::Error;
 use std::io;
 use std::path::{Path, PathBuf};
 use std::sync::TryLockError;
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum SnapshotError {
+    IncorrectSnapshotRequested(CryptoHash, CryptoHash),
+    SnapshotNotFound,
+    LockWouldBlock,
+    Other(String),
+}
+
+impl std::fmt::Display for SnapshotError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SnapshotError::IncorrectSnapshotRequested(requested, available) => write!(
+                f,
+                "Wrong state snapshot. Requested: {:?}, Available: {:?}",
+                requested, available
+            ),
+            SnapshotError::SnapshotNotFound => write!(f, "No state snapshot available"),
+            SnapshotError::LockWouldBlock => {
+                write!(f, "Accessing state snapshot would block. Retry in a few seconds.")
```

Comment:
For my education, under what circumstances may that happen? Is that resource starvation when competing with state sync? Can we give resharding priority?

Reply:
State sync takes a lock on the snapshot in two scenarios, to generate a snapshot and to compact it. Both these scenarios can be potentially time consuming and resharding first needs to wait for these operations to complete.

I don't think there should be a starvation problem here as we generate snapshots only once per epoch.

Reply:
No lock on compaction.